beartype/typing.py

```diff
@@ -288,3 +288,211 @@
+         ``@runtime_checkable`` is to inherit from
+         :class:`beartype.typing.Protocol` instead:
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+        """
+
+        _abc_inst_check_cache: Dict[Type, bool]
+
+        def __new__(
+            mcls: Type[_TT],
+            name: str,
+            bases: Tuple[Type, ...],
+            namespace: Dict[str, Any],
+            **kw: Any,
+        ) -> _TT:
+            # See <https://github.com/python/mypy/issues/9282>
+            cls = super().__new__(mcls, name, bases, namespace, **kw)  # type: ignore [misc]
+            # Prefixing this class member with "_abc_" is necessary to prevent
+            # it from being considered part of the Protocol. (See
+            # <https://github.com/python/cpython/blob/main/Lib/typing.py>.)
+            cls._abc_inst_check_cache = {}
+
+            return cls
+
+        def __instancecheck__(cls, inst: Any) -> bool:
+            try:
+                # This has to stay *super* tight! Even adding a mere assertion
+                # can add ~50% to the best case runtime!
+                return cls._abc_inst_check_cache[type(inst)]
+            except KeyError:
+                # If you're going to do *anything*, do it here. Try not to
+                # expand the rest of this method if you can avoid it.
+                inst_t = type(inst)
+                bases_pass_muster = True
+
+                for base in cls.__bases__:
+                    if base is cls or base.__name__ in (
+                        "Protocol",
+                        "Generic",
+                        "object",
+                    ):
+                        continue
+
+                    if not isinstance(inst, base):
+                        bases_pass_muster = False
+                        break
+
+                cls._abc_inst_check_cache[
+                    inst_t
+                ] = bases_pass_muster and cls._check_only_my_attrs(inst)
+
+                return cls._abc_inst_check_cache[inst_t]
+
+        def _check_only_my_attrs(cls, inst: Any) -> bool:
+            attrs = set(cls.__dict__)
+            attrs.update(cls.__dict__.get("__annotations__", {}))
+            attrs.intersection_update(_get_protocol_attrs(cls))
+
+            for attr in attrs:
+                if not hasattr(inst, attr):
+                    return False
+                elif callable(getattr(cls, attr, None)) and getattr(inst, attr) is None:
+                    return False
+
+            return True
+
+    @runtime_checkable
+    class Protocol(_Protocol, Generic, metaclass=_CachingProtocolMeta):  # type: ignore [no-redef]
```

Comment:
Just a heads up: I don't think deriving from `Generic` here does anything. It was an experiment to see if I could work through the issue I raised in [this comment](https://github.com/beartype/beartype/pull/86#pullrequestreview-860250362).

Reply:
Removed.